velox/expression/tests/ArgGeneratorTest.cpp

```diff
@@ -0,0 +1,160 @@
+          .integerVariable("r_scale", "min(scale + 1, 18)")
+          .returnType("decimal(r_precision, r_scale)")
+          .argumentType("decimal(precision, scale)")
+          .build();
+
+  std::function<void(const std::vector<TypePtr>&, const TypePtr&)> verify =
+      [](const std::vector<TypePtr>& argumentTypes, const TypePtr& returnType) {
+        EXPECT_EQ(argumentTypes.size(), 1);
+        const auto [precision, scale] =
+            getDecimalPrecisionScale(*argumentTypes[0]);
+        const auto [rPrecision, rScale] = getDecimalPrecisionScale(*returnType);
+        ASSERT_EQ(rPrecision, std::min(38, precision + scale + 1));
+        ASSERT_EQ(rScale, std::min(scale + 1, 18));
+      };
+
+  const auto generator = std::make_shared<UnaryArgGenerator>();
+  for (auto returnType : {DECIMAL(10, 2), DECIMAL(38, 18)}) {
+    SCOPED_TRACE(fmt::format("returnType: {}", returnType->toString()));
+    auto argTypes = generateArgumentTypes(generator, *signature, returnType);
+    verify(argTypes, returnType);
+  }
+
+  auto argTypes = generateArgumentTypes(generator, *signature, DECIMAL(38, 20));
+  EXPECT_EQ(argTypes.size(), 0);
+}
+
+TEST_F(ArgGeneratorTest, binary) {
+  auto signature =
+      exec::FunctionSignatureBuilder()
+          .integerVariable("a_scale")
+          .integerVariable("b_scale")
+          .integerVariable("a_precision")
+          .integerVariable("b_precision")
+          .integerVariable(
+              "r_precision",
+              "min(38, max(a_precision - a_scale, b_precision - b_scale) + max(a_scale, b_scale) + 1)")
+          .integerVariable("r_scale", "max(a_scale, b_scale)")
+          .returnType("decimal(r_precision, r_scale)")
+          .argumentType("decimal(a_precision, a_scale)")
+          .argumentType("decimal(b_precision, b_scale)")
+          .build();
+
+  std::function<void(const std::vector<TypePtr>&, const TypePtr&)> verify =
+      [](const std::vector<TypePtr>& argumentTypes, const TypePtr& returnType) {
+        EXPECT_EQ(argumentTypes.size(), 2);
+        const auto [aPrecision, aScale] =
+            getDecimalPrecisionScale(*argumentTypes[0]);
+        const auto [bPrecision, bScale] =
+            getDecimalPrecisionScale(*argumentTypes[1]);
+        const auto [rPrecision, rScale] = getDecimalPrecisionScale(*returnType);
+        ASSERT_EQ(
+            rPrecision,
+            std::min(
+                38,
+                std::max(aPrecision - aScale, bPrecision - bScale) +
+                    std::max(aScale, bScale) + 1));
+        ASSERT_EQ(rScale, std::max(aScale, bScale));
+      };
+
+  const auto generator = std::make_shared<BinaryArgGenerator>();
+  for (auto returnType :
+       {DECIMAL(10, 2), DECIMAL(38, 20), DECIMAL(38, 38), DECIMAL(38, 0)}) {
+    SCOPED_TRACE(fmt::format("returnType: {}", returnType->toString()));
+    auto argTypes = generateArgumentTypes(generator, *signature, returnType);
+    verify(argTypes, returnType);
```

Comment:
I'm thinking that we can write a more generic verify function that uses SignatureBinder to resolve return type from the inputs given by 'generator' and assert that this return type matches the original one. What do you think?

Reply:
Added resolveReturnType function to resolve return type from the inputs given by generator.